aggregator/src/aggregator/aggregation_job_creator.rs

```diff
@@ -347,6 +380,164 @@ impl<C: Clock + 'static> AggregationJobCreator<C> {
+                        //    batch based on already-existing aggregation jobs; otherwise, we risk
+                        //    aggregating more than max_batch_size reports together.
+                        // Choose the maximal size meeting all of these requirements.
+                        let aggregation_job_size = [
+                            unaggregated_report_ids.len(),
+                            this.max_aggregation_job_size,
+                            max_batch_size - max_size,
+                        ]
+                        .into_iter()
+                        .min()
+                        .unwrap();
+
+                        if aggregation_job_size < this.min_aggregation_job_size {
+                            if !is_batch_new
+                                && max_size < min_batch_size
+                                && max_size + aggregation_job_size >= min_batch_size
+                            {
+                                // This batch is short of the minimum batch size, and requires an
+                                // unusually small aggregation job (smaller than the normal minimum
+                                // aggregation job size) for it to be ever completed. Go ahead and
+                                // generate one. (We also wait until the size of the aggregation job
+                                // we can generate will meet the minimum configured batch size, in
+                                // an attempt to minimize the number of "small" aggregation jobs we
+                                // create.)
+                            } else if !is_batch_new {
+                                // Move on to the next unfilled batch to see if we can allocate
+                                // reports to it.
+                                (is_batch_new, batch) = batch_iter.next().unwrap();
+                                max_size = batch.max_size();
+                                continue;
+                            } else {
+                                // We have run out of preexisting batches to evaluate adding reports
+                                // to. Trying additional new batches won't help (since all of the
+                                // relevant parameters will be the same as in this iteration), so
+                                // stop generating aggregation jobs.
+                                break;
+                            }
+                        }
+
+                        // Generate an aggregation job, then update batch metadata & continue.
+                        let aggregation_job_id = random();
+                        debug!(
+                            task_id = %task.id(),
+                            batch_id = %batch.id(),
+                            %aggregation_job_id,
+                            report_count = aggregation_job_size,
+                            "Creating aggregation job"
+                        );
```

Comment:
Have we plumbed a metrics collector into this struct? If so, we should bump a counter for aggregation job creation. If not, let's put a TODO comment and file an issue about complementing `debug!` and `info!` prints with metrics.

Reply:
Filed https://github.com/divviup/janus/issues/680.